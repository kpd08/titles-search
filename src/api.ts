import axios from 'axios';
import { API_KEY, API_URL } from './constants';

export const fetchProductions = async (
  searchQuery: string,
  page: number = 1
): Promise<{ productions: Array<Production>; totalResults: number }> => {
  try {
    const { data } = await axios.get<SearchResponse<Production>>(
      `${API_URL}?apikey=${API_KEY}&s=${encodeURIComponent(
        searchQuery
      )}&page=${page}`
    );
    if (data.Error) {
      throw new Error(data.Error);
    }
    return {
      productions: data.Search || [],
      totalResults: Number.parseInt(data.totalResults!, 10) || 0,
    };
  } catch (error) {
    throw new Error(error);
  }
};
