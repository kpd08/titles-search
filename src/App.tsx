import React, { useState, useCallback, useMemo } from 'react';
import { fetchProductions as apiFetchProductions } from './api';
import { Layout } from './components/layout';
import { LoaderOverlay } from './components/loader';
import { ProductionEntry } from './components/production-entry';
import { SearchBox } from './components/search-box';
import debounce from 'lodash/debounce';
import { ErrorBox } from './components/error-box';
import { FALLBACK_ERROR_MSG } from './constants';

export default function App() {
  const [productions, setProductions] = useState<Array<Production>>([]);
  const [searchValue, setSearchValue] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState<string | null>(null);

  const fetchProductions = async (searchValue: string) => {
    try {
      setErrorMsg(null);
      if (searchValue.length < 3) {
        setProductions([]);
        return;
      }
      setIsLoading(true);
      const { productions } = await apiFetchProductions(searchValue);
      setProductions(productions);
    } catch (error) {
      const errorMessage = error?.message || FALLBACK_ERROR_MSG;
      setErrorMsg(errorMessage);
      setProductions([]);
    } finally {
      setIsLoading(false);
    }
  };

  const debouncedFetchProductions = useCallback(
    debounce(fetchProductions, 500),
    []
  );

  const handleSearchBoxChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const { value } = event.target;
    setSearchValue(value);
    debouncedFetchProductions(value);
  };

  const hasResults = useMemo(() => !!(productions && productions.length), [
    productions,
  ]);

  return (
    <Layout>
      <div className="flex flex-col">
        <SearchBox
          placeholder="Search titles"
          value={searchValue}
          onChange={handleSearchBoxChange}
        />
        {isLoading && <LoaderOverlay />}
        {errorMsg && <ErrorBox>{errorMsg}</ErrorBox>}
        {!hasResults && !isLoading && !errorMsg && <h1>No titles</h1>}
        {hasResults &&
          !isLoading &&
          productions.map((production) => (
            <ProductionEntry production={production} key={production.imdbID} />
          ))}
      </div>
    </Layout>
  );
}
