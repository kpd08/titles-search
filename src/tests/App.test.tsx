import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import * as api from '../api';
import App from '../App';

jest.mock('../api');

const mockedFetchProductions = api.fetchProductions as jest.Mock;

const mockedProduction: Production = {
  Poster: 'some.poster.url',
  Title: 'Some title',
  Type: 'movie',
  Year: '2018',
  imdbID: '111',
};

describe('App', () => {
  beforeEach(() => {
    mockedFetchProductions.mockClear().mockResolvedValue({
      productions: [mockedProduction],
      totalResults: 0,
    });
  });

  test('it renders as expected', () => {
    render(<App />);
  });

  test('it shows `no titles` if there are no entries', () => {
    const { queryByText } = render(<App />);
    expect(queryByText(/no titles/i)).toBeInTheDocument();
  });

  test('show/hide loader overlay when fetching data', async () => {
    const { getByTestId } = render(<App />);
    userEvent.type(getByTestId('search-box'), 'test');
    await waitFor(() => {
      expect(getByTestId('loader-overlay')).toBeInTheDocument();
    });
  });

  test('shows error if something went wrong', async () => {
    mockedFetchProductions.mockRejectedValue(new Error('some error'));
    const { getByTestId } = render(<App />);
    userEvent.type(getByTestId('search-box'), 'test');
    await waitFor(() => {
      expect(mockedFetchProductions).toHaveBeenCalledWith('test');
      expect(mockedFetchProductions).toHaveBeenCalledTimes(1);
      expect(getByTestId('error-box')).toHaveTextContent('some error');
    });
  });

  test('shows list of results', async () => {
    const { getByTestId, getAllByTestId } = render(<App />);
    userEvent.type(getByTestId('search-box'), 'test');
    await waitFor(() => {
      expect(getAllByTestId('production-details')).toMatchSnapshot();
      expect(mockedFetchProductions).toHaveBeenCalledWith('test');
      expect(mockedFetchProductions).toHaveBeenCalledTimes(1);
    });
  });
});
