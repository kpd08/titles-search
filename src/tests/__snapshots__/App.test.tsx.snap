// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App shows list of results 1`] = `
Array [
  <div
    class="flex flex-col items-center md:flex-row md:items-start my-2 p-4 border-b-2 border-gray-400 border-opacity-50"
    data-testid="production-details"
  >
    <img
      alt="Some title"
      class="object-cover rounded-full justify-self-center w-40 h-40 md:h-56 md:rounded-md shadow-lg md:mx-4"
      src="some.poster.url"
    />
    <div
      class="flex flex-col items-center md:items-start my-4 md:my-0"
    >
      <h1
        class="text-lg md:text-3xl text-center"
      >
        Some title
      </h1>
      <span
        class="text-gray-600"
      >
        2018
      </span>
    </div>
  </div>,
]
`;
